smack-experimental/src/main/java/org/jivesoftware/smackx/message_fastening/provider/FasteningElementProvider.java

```diff
@@ -0,0 +1,83 @@
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.message_fastening.provider;
+
+import java.io.IOException;
+
+import org.jivesoftware.smack.packet.ExtensionElement;
+import org.jivesoftware.smack.packet.XmlEnvironment;
+import org.jivesoftware.smack.parsing.SmackParsingException;
+import org.jivesoftware.smack.provider.ExtensionElementProvider;
+import org.jivesoftware.smack.util.PacketParserUtils;
+import org.jivesoftware.smack.util.ParserUtils;
+import org.jivesoftware.smack.xml.XmlPullParser;
+import org.jivesoftware.smack.xml.XmlPullParserException;
+import org.jivesoftware.smackx.message_fastening.MessageFasteningManager;
+import org.jivesoftware.smackx.message_fastening.element.ExternalElement;
+import org.jivesoftware.smackx.message_fastening.element.FasteningElement;
+
+public class FasteningElementProvider extends ExtensionElementProvider<FasteningElement> {
+
+    public static final FasteningElementProvider TEST_INSTANCE = new FasteningElementProvider();
+
+    @Override
+    public FasteningElement parse(XmlPullParser parser, int initialDepth, XmlEnvironment xmlEnvironment) throws XmlPullParserException, IOException, SmackParsingException {
+        FasteningElement.Builder builder = FasteningElement.builder();
+        builder.setOriginId(parser.getAttributeValue("", FasteningElement.ATTR_ID));
+        if (ParserUtils.getBooleanAttribute(parser, FasteningElement.ATTR_CLEAR, false)) {
+            builder.setClear();
+        }
+        if (ParserUtils.getBooleanAttribute(parser, FasteningElement.ATTR_SHELL, false)) {
+            builder.setShell();
+        }
+
+        outerloop: while (true) {
+            XmlPullParser.Event tag = parser.next();
+            switch (tag) {
+                case START_ELEMENT:
+                    String name = parser.getName();
+                    String namespace = parser.getNamespace();
+
+                    // Parse external payload
+                    if (MessageFasteningManager.NAMESPACE.equals(namespace) && ExternalElement.ELEMENT.equals(name)) {
+                        ExternalElement external = new ExternalElement(
+                                parser.getAttributeValue("", ExternalElement.ATTR_NAME),
+                                parser.getAttributeValue("", ExternalElement.ATTR_ELEMENT_NAMESPACE));
+                        builder.addExternalPayload(external);
+                        continue;
+                    }
+
+                    // Parse wrapped payload
+                    ExtensionElement wrappedPayload = PacketParserUtils.parseExtensionElement(name, namespace, parser, xmlEnvironment);
+                    builder.addWrappedPayload(wrappedPayload);
+                    break;
+
+                case END_ELEMENT:
+                    name = parser.getName();
+                    namespace = parser.getNamespace();
+
+                    if (FasteningElement.ELEMENT.equals(name) && MessageFasteningManager.NAMESPACE.equals(namespace)) {
```

Comment:
Strictly speaking, that is not robust, as it would also break the outerloop if there is a fastening element within a fastening element. I know, this is very unlikely, but the correct idiom is to break if the depth equals the initial depth, which is given as argument to the parse method.

In order to avoid other patters being used in places where it potentially matters me, we should simply use this one pattern everywhere.